hdfs-scheduler/src/main/java/org/apache/mesos/hdfs/scheduler/Reconciler.java

```diff
@@ -0,0 +1,110 @@
+package org.apache.mesos.hdfs.scheduler;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.mesos.hdfs.state.IPersistentStateStore;
+import org.apache.mesos.Protos.TaskState;
+import org.apache.mesos.Protos.TaskStatus;
+import org.apache.mesos.Protos;
+import org.apache.mesos.SchedulerDriver;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Observable;
+import java.util.Observer;
+import java.util.Set;
+
+/**
+ * HDFS Mesos Framework Reconciler class implementation.
+ */
+public class Reconciler implements Observer {
+  private final Log log = LogFactory.getLog(HdfsScheduler.class);
+  private IPersistentStateStore store;
+  private Set<String> pendingTasks;
+
+  public Reconciler(IPersistentStateStore pStore) {
+    store = pStore;
+    pendingTasks = new HashSet<String>();
+  }
+
+  public void reconcile(SchedulerDriver driver) {
+    pendingTasks = store.getAllTaskIds();
+    logPendingTasks();
+    explicitlyReconcileTasks(driver);
+    implicitlyReconcileTasks(driver);
+  }
+
+  public void update(Observable obs, Object obj) {
+    TaskStatus status = (TaskStatus) obj;
+
+    String taskId = status.getTaskId().getValue();
+    log.info("Received task update for: " + taskId);
+
+    if (!complete()) {
+      log.info("Reconciliation is NOT complete");
+
+      if (taskIsPending(taskId)) {
+        log.info(String.format("Reconciling Task '%s'.", taskId));
+        pendingTasks.remove(taskId);
+      } else {
```

Comment:
It wasn't necessarily already reconciled.  This may be something else altogether.  It could potentially be a notification about the Status of a Task which we did not persist information about.  All I know for sure is that it's a Task that doesn't address any of the Tasks pending reconciliation.